src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -365,25 +365,167 @@ public class Options
+
+            // Read until we either get requested data or we hit end of stream.
+            public async Task<int> ReadBlockAsync(byte[] buffer, int offset, int size)
+            {
+                int totalLength = 0;
+
+                while (size != 0)
+                {
+                    int readLength = await ReadAsync(buffer, offset, size).ConfigureAwait(false);
+
+                    if (readLength == 0)
+                    {
+                        break;
+                    }
+
+                    totalLength += readLength;
+                    offset += readLength;
+                    size -= readLength;
+                }
+
+                return totalLength;
+            }
+
+            public async Task<int> ReadBlockAsync(char[]  result, int offset, int size)
+            {
+                byte[] buffer = new byte[size];
+                int readLength = await ReadBlockAsync(buffer, 0, size);
+
+                string asString = System.Text.Encoding.ASCII.GetString(buffer, 0, readLength);
+
+                for (int i = 0; i < readLength; i++)
+                {
+                    result[offset + i ] = asString[i];
+                }
+
+                return readLength;
+            }
+
+            public async Task<string> ReadToEndAsync()
+            {
+                byte[] buffer = new byte[_blockSize];
+                int offset = 0;
+                int totalLength = 0;
+                int bytesRead;
+
+                do
+                {
+                    bytesRead = await ReadAsync(buffer, offset, buffer.Length - offset);
+                    totalLength += bytesRead;
+                    offset+=bytesRead;
+
+                    if (bytesRead == buffer.Length)
+                    {
+                        byte[] newBuffer = new byte[buffer.Length + _blockSize];
+                        buffer.CopyTo(newBuffer, 0);
+                        offset = buffer.Length;
+                        buffer = newBuffer;
+                    }
+                } while (bytesRead > 0);
+
+                return System.Text.Encoding.ASCII.GetString(buffer, 0, totalLength);
+            }
+
+            public string ReadLine()
+            {
+                return ReadLineAsync().GetAwaiter().GetResult();
+            }
+
+            public async Task<string> ReadLineAsync()
+            {
+                int index = 0;
+                int startSearch = _start;
+
+                while (true)
+                {
+                    if (_start == _end || index == -1)
+                    {
+                        // We either have no data or we did not find LF in stream.
+                        // In either case, read more.
+                        if (_end + 2 > _buffer.Length)
+                        {
+                            // We no longer have space to read CRLF. Allocate new buffer and start over.
+                            byte[] newBuffer = new byte[_buffer.Length + _blockSize];
```

Comment:
It might simplify things a bit to use ArrayBuffer from the product here.

Reply:
I'll do it as separate change. I assume I should move it to common place to avoid direct linking like we do in UnitTests, right?